### DISCUSSION TITLE
Have you guys heard about LFortran?

### DISCUSSION THREAD

**Discussion by gabrielfougeron** [pyodide/pyodide#3516] [Ideas]
Dear Pyodide team,

I learnt today about [LFortran](https://github.com/lfortran/lfortran), a "Modern interactive LLVM-based Fortran compiler". This project is not mature (alpha stage), but specifically targets WebAssembly as a platform.

Were you guys aware of the existence of this project?
I've come to understand that fortran code in blas, lapack and scipy has been a source of problems for pyodide. Could LFortran help in that regard?

Best,

**Comment by rth** [comment]
Yes, we are hoping we would be able to use lfortran in the future. For now, it's not able to compile scipy (although they are working on it), so we cannot use it yet.

**Comment by hoodmane** [comment]
@certik what is the status of things over there? I see https://github.com/lfortran/lfortran/issues/550 is still open, maybe I should come back and try to investigate that further...

**Comment by certik** [comment]
We are finishing our work with Minpack, one of the Fortran packages that SciPy uses. We can compile it and run several examples correctly, but we want to run all examples. We can also parse all of Fortran in SciPy into AST. Once Minpack is fully working, it shouldn't be that hard to get the rest of scipy working as well, we will look into that once Minpack is done.

We also have a large effort to compile the Fortran Package Manager (fpm) which is using a lot of modern Fortran features (pretty much all of them...) and we are now at 77% of compiling it to our internal representation. After we get to 100%, we will work on compiling our intermediate representation to LLVM for all of fpm. I don't know if you care about modern Fortran or only Scipy's Fortran.

Regarding WASM, we have two backends, the main one is LLVM, and LLVM can produce WASM. You can help us making the LLVM->WASM route more smooth for the end user. Many people can help with that, so we are focusing on the actual compilation (semantics). We also have a direct WASM backend, the advantage of that one is that it does not require LLVM, so it can run in the browser with a small download and it is fast to compile, here is a demo: https://dev.lfortran.org/.

If anyone of you is interested to help out, please get in touch, we could always use more hands.

**Reply by hoodmane** [reply]
> I don't know if you care about modern Fortran or only Scipy's Fortran.

We are primarily concerned with Scipy since so many people use it. It will be a tremendous help for us once that works. (Well and we also need to be able to make dynamic libraries.)  Probably some people out there have Python packages written using modern Fortran so that can't hurt.

**Reply by certik** [reply]
Yes, we will get LFortran compiling all of SciPy. Once Minpack is solid, we will write a blog post documenting our current status, and we will look at what is needed for the rest of SciPy.

I was thinking a nice workflow could be to just get LFortran to be used for Minpack inside SciPy and see if everything works, or there are some integration issues. Then we'll tackle another package and one by one we get all of SciPy working.

**Comment by certik** [comment]
So we got Minpack working: https://fortran-lang.discourse.group/t/lets-get-lfortran-to-compile-minpack/4550/5. We'll have a look soon what else is needed for the rest of SciPy.

Is anybody here interested in testing out LFortran with just Minpack in SciPy? Use the latest master (main). There might be some bugs in integration, if you report them, we'll fix them. Minpack itself (sources copied from SciPy) seems to work.
